### ISSUE TITLE
nbgrader doesn't work when jupyterhub has SSL enabled.

### ISSUE THREAD

**Issue by jonathanmorgan** [jupyter/nbgrader#400]
We're having a problem with getting formgrade to work with jupyterhub when jupyterhub has SSL enabled.  When I turn SSL on in jupyterhub by specifying an ssl_key and ssl_cert, jupyterhub is fine, but the nbgrader formgrade application becomes inaccessible.  I am not sure what is going on, but nothing shows up in the formgrade logs, and I'm not sure where to find logging from the proxy server.  What you see in the browser when you hit 'https://<domain>/hub/nbgrader/<classname>" is ongoing loading, like what you see when you visit a page using HTTP that is only available in HTTPS.  It doesn't time out quick, it doesn't get an error, it just keeps spinning, and it can go for minutes.

I think this is similar to the problem outlined in issue 178 ( [https://github.com/jupyter/nbgrader/pull/178](https://github.com/jupyter/nbgrader/pull/178) ) - there, in file nbgrader/auth/hubauth.py, the domain needed to be configurable in the auth URL.  I think we also need to be able to configure the protocol string at the beginning, so it can be "https".

In addition, I expect we'll also want this to carry through to the rendering of the URLs inside the formgrade application pages, as well, so they also use the same protocol string.

**Comment by jhamrick** [comment]
Ok, can you try customizing the `hub_base_url` so that it uses https -- e.g. `c.HubAuth.hub_base_url = "https://<domain>:8000"`? I am not sure if that will fix it in every case, but it should hopefully fix some of the issues you're having...

**Comment by jonathanmorgan** [comment]
That didn't work.  In looking at nbgrader/auth/hubauth.py, it appears to be appending "http://" to the front of the URLs throughout:

    43    def _hub_base_url_default(self):
    44        return 'http://{}:{}'.format(self.hub_address, self.hub_port)
    ...
    96        self._hubapi_base_url = 'http://{}:{}'.format(self.hubapi_address, self.hubapi_port)
    97        self._proxy_base_url = 'http://{}:{}'.format(self.proxy_address, self.proxy_port)
    ...
    105        target = 'http://{}:{}'.format(ip, self._port)

**Comment by jonathanmorgan** [comment]
Also, if you add a configuration parameter for protocol (not sure how you'd prefer to do that, could actually make it a string that defaults to "http", or could have an SSL flag), you'll also likely want to have a parameter to allow nbgrader to be configured to either not verify the certificate (for simple self-signed certificates), or to accept a custom trust store (so you could specify the certificate's pem as the trust store, and at least verify that it is the self-signed certificate you are expecting).

I think I saw in the source that you use the python requests library for connections.  If so, here is a stack-overflow post with discussion on how you do both of these with requests: [http://stackoverflow.com/questions/10667960/python-requests-throwing-up-sslerror](http://stackoverflow.com/questions/10667960/python-requests-throwing-up-sslerror).  Other libraries should have something similar.

Depending on your timing, I'd be willing to help with this, but I'd need an introduction to the code if I am to do it efficiently.

**Comment by jonathanmorgan** [comment]
Better, here is the actual documentation: [http://docs.python-requests.org/en/latest/user/advanced/#ssl-cert-verification](http://docs.python-requests.org/en/latest/user/advanced/#ssl-cert-verification)

**Comment by willingc** [comment]
@jhamrick @jonathanmorgan Just an FYI...once you decide your approach. If you have questions for the requests core devs, all three are friends of mine and will be super helpful via IRC.

**Comment by jhamrick** [comment]
Hmm, because nbgrader is running behind JupyterHub, it shouldn't need to deal with anything relating to SSL itself, I think. In terms of the examples of `http` you listed:

```
43    def _hub_base_url_default(self):
44        return 'http://{}:{}'.format(self.hub_address, self.hub_port)
```

This gets overwritten if you set `hub_base_url` in your config (this is just the default value).

```
96        self._hubapi_base_url = 'http://{}:{}'.format(self.hubapi_address, self.hubapi_port)
97        self._proxy_base_url = 'http://{}:{}'.format(self.proxy_address, self.proxy_port)
```

These shouldn't be https I think because the proxy and hub api aren't actually using SSL -- they're just listening on localhost on different ports.

```
105        target = 'http://{}:{}'.format(ip, self._port)
```

That should also be fine, because the SSL portion is through JupyterHub, not through nbgrader.

The SSL for my setup is through a nginx proxy in front of JupyterHub, so that works a little differently, and I haven't setup JupyterHub to actually use SSL with nbgrader. I can try to set up a toy example and see if I can get it to work, though. In the meantime, maybe @jdfreder knows the answer, since I think he set up SSL+JupyterHub+nbgrader for Brian's class?

@willingc thanks! I'm not sure if this will actually require us to do anything with requests, but if it does I'll let you know.

**Comment by jhamrick** [comment]
Hmm, ok, so I created a simple setup with SSL and it seems to work for me. Here's my `jupyterhub_config.py`:

```
c = get_config()
c.JupyterHub.ssl_cert = './jupyterhub_cert.pem'
c.JupyterHub.ssl_key = './jupyterhub_key.pem'
c.JupyterHub.proxy_auth_token = 'foo'
c.Authenticator.admin_users = ['jhamrick']
c.Authenticator.whitelist = ['jhamrick']
```

and my `nbgrader_config.py`:

```
c = get_config()
c.NbGrader.course_id = "example_course"
c.TransferApp.exchange_directory = "/tmp/exchange"
c.FormgradeApp.ip = "127.0.0.1"
c.FormgradeApp.port = 9000
c.FormgradeApp.authenticator_class = "nbgrader.auth.hubauth.HubAuth"
c.HubAuth.hub_base_url = "https://localhost:8000"
c.HubAuth.notebook_url_prefix = "path/to/class_files"
c.HubAuth.graders = ["jhamrick"]
c.HubAuth.generate_hubapi_token = True
c.HubAuth.hub_db = "/tmp/jupyterhub/jupyterhub.sqlite"
```

Can you let me know what is different between your config and this config?

**Comment by jonathanmorgan** [comment]
I will look now.  Thanks again for your help.

Jon

-- 

"The man with the new idea is a Crank until the idea succeeds."
- Mark Twain, from 'Following the Equator: A Journey Around the World'

**Comment by jonathanmorgan** [comment]
I configured both my server and the production server exactly as this example is configured, except with appropriate paths for where things are located on each server.  The only difference is that my server’s certificate is self-signed, the production one is legit.

On each, after I turn on SSL and verify that the jupyterhub works just fine, I then reference the form grade URL:

https://<domain>:8000/hub/nbgrader/2015-fall-big_data/ <https://bigdataforasocialscience.com:8000/hub/nbgrader/2015-fall-big_data/>

In the form grade log, I see this:

	127.0.0.1 - - [01/Sep/2015 23:52:22] "GET /hub/nbgrader/2015-fall-big_data/ HTTP/1.1" 302 -

Then, in the log output for jupyterhub, I see this:

	[I 2015-09-01 23:52:22.260 JupyterHub log:100] 200 GET /hub/api/authorizations/cookie/jupyter-hub-token/[secret] (instructor@127.0.0.1) 11.81ms

And then regardless of server, the browser either:

redirects to http://<domain>:8000/hub/nbgrader/2015-fall-big_data/assignments/ <http://<domain>:8000/hub/nbgrader/2015-fall-big_data/assignments/> (HTTP now, not HTTPS) and stops responding if c.HubAuth.hub_base_url = “https://<domain>:8000”.
or just stops responding if I have c.HubAuth.hub_base_url = “https://localhost:8000”.

I think I need to have the c.HubAuth.hub_base_url set so it includes the actual domain name, not localhost, but I am not sure.  I had to change it to just the actual domain name, with no port, to get it to work without SSL (Are you running the browser on the computer that serves the jupyterhub?).  This is originally what led me to look at places where URLs are created - without setting c.HubAuth.hub_base_url, the URLs in the rendered HTML referenced localhost or 127.0.0.1, rather than the server.

But, it doesn’t work either way.  Might it be a problem with the way the proxy is configured?

-- 

"The man with the new idea is a Crank until the idea succeeds."
- Mark Twain, from 'Following the Equator: A Journey Around the World'

**Comment by jonathanmorgan** [comment]
Not sure if this helps, but I looked more closely at the redirects, and here is what happens:

https://<domain>:8000/hub/nbgrader/2015-fall-big_data/ <https://<domain>:8000/hub/nbgrader/2015-fall-big_data/>
https://<domain>:8000/hub/login?next=/hub/nbgrader/2015-fall-big_data <https://<domain>:8000/hub/login?next=/hub/nbgrader/2015-fall-big_data>
if not authenticated, you log in, then: https://<domain>:8000/hub/login?next=%2Fhub%2Fnbgrader%2F2015-fall-big_data <https://data.jrn.cas.msu.edu:8000/hub/login?next=/hub/nbgrader/2015-fall-big_data>
At this point, regardless of whether you were logged in or not, the login page knows you are logged in, so it redirects.
Browser stops responding.  In Chrome, you end up at "http://<domain>:8000/hub/nbgrader/2015-fall-big_data/ <http://<domain>:8000/hub/nbgrader/2015-fall-big_data/>assignments”, so I think it sends you to “https://<domain>:8000/hub/nbgrader/2015-fall-big_data/ <https://<domain>:8000/hub/nbgrader/2015-fall-big_data/>“ then redirects you to "http://<domain>:8000/hub/nbgrader/2015-fall-big_data/ <http://<domain>:8000/hub/nbgrader/2015-fall-big_data/>assignments/”.  But I’m not sure.  I can’t figure out how to get visibility into the chain of requests occurring.

If it would help, I can email you the exact configuration files and leave my test server set up this way and send you the URLs so you can see what is going on.  I’m hesitant to put them in posts here.

Jon

-- 

"The man with the new idea is a Crank until the idea succeeds."
- Mark Twain, from 'Following the Equator: A Journey Around the World'

**Comment by jhamrick** [comment]
Ahh, I think this might be a difference between 0.2.1 and master. I realized before I tried to run this on master, but you're probably running from 0.2.1, and sure enough, it doesn't seem to work on 0.2.1. I bet it's because of a difference between Flask and tornado.

I will try to figure out the specific fix for Flask and merge into the 0.2.x branch. I am planning on doing a release of 0.2.2 shortly anyway and will make sure this makes it in before that.

**Comment by jonathanmorgan** [comment]
Yup, we are on 0.2.1, from pip.  I think everyone is fine running sans SSL for a bit.

Jon

-- 

"The man with the new idea is a Crank until the idea succeeds."
- Mark Twain, from 'Following the Equator: A Journey Around the World'

**Comment by jhamrick** [comment]
Closed by #410

**Comment by jonathanmorgan** [comment]
works fine in 0.2.2.
